src/hotspot/share/opto/addnode.cpp

```diff
@@ -396,9 +396,182 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
+  if (mul.is_valid_with(pattern)) {
+    return mul;
+  }
+
+  // We've tried everything.
+  return Multiplication::make_invalid();
+}
+
+// Try to match `n = a + a`. On success, return a struct with `.valid = true`, `variable = a`, and `multiplier = 2`.
+// The method matches `n` for pattern: a + a.
+AddNode::Multiplication AddNode::find_simple_addition_pattern(const Node* n, BasicType bt) {
+  if (n->Opcode() == Op_Add(bt) && n->in(1) == n->in(2)) {
+    return Multiplication{true, n->in(1), 2};
+  }
+
+  return Multiplication::make_invalid();
+}
+
+// Try to match `n = a << CON`. On success, return a struct with `.valid = true`, `variable = a`, and
+// `multiplier = 1 << CON`.
+// Match `n` for pattern: a << CON.
+// Note that the power-of-2 multiplication optimization could potentially convert a MulNode to this pattern.
+AddNode::Multiplication AddNode::find_simple_lshift_pattern(const Node* n, BasicType bt) {
+  // Note that power-of-2 multiplication optimization could potentially convert a MulNode to this pattern
+  if (n->Opcode() == Op_LShift(bt) && n->in(2)->is_Con()) {
+    Node* con = n->in(2);
+    if (!con->is_top()) {
+      return Multiplication{true, n->in(1), java_shift_left(1, con->get_int(), bt)};
+    }
+  }
+
+  return Multiplication::make_invalid();
+}
+
+// Try to match `n = CON * a`. On success, return a struct with `.valid = true`, `variable = a`, and `multiplier = CON`.
+// Match `n` for patterns:
+//     - (1) CON * a
+//     - (2) a * CON
+AddNode::Multiplication AddNode::find_simple_multiplication_pattern(const Node* n, BasicType bt) {
+  // This optimization technically only produces MulNode(CON, a), but we might as match MulNode(a, CON), too.
+  if (n->Opcode() == Op_Mul(bt) && (n->in(1)->is_Con() || n->in(2)->is_Con())) {
+    // Pattern (1)
+    Node* con = n->in(1);
+    Node* base = n->in(2);
+
+    // Pattern (2)
+    if (!con->is_Con()) {
+      // swap ConNode to lhs for easier matching
+      swap(con, base);
+    }
+
+    if (!con->is_top()) {
+      return Multiplication{true, base, con->get_integer_as_long(bt)};
+    }
+  }
+
+  return Multiplication::make_invalid();
+}
+
+// Try to match `n = (a << CON1) + (a << CON2)`. On success, return a struct with `.valid = true`, `variable = a`, and
+// `multiplier = (1 << CON1) + (1 << CON2)`.
+// Match `n` for patterns:
+//     - (1) (a << CON) + (a << CON)
+//     - (2) (a << CON) + a
+//     - (3) a + (a << CON)
+//     - (4) a + a
+// Note that one or both of the term of the addition could simply be `a` (i.e., a << 0) as in pattern (4).
+AddNode::Multiplication AddNode::find_power_of_two_addition_pattern(const Node* n, BasicType bt) {
+  if (n->Opcode() == Op_Add(bt) && n->in(1) != n->in(2)) {
+    Multiplication lhs = find_simple_lshift_pattern(n->in(1), bt);
+    Multiplication rhs = find_simple_lshift_pattern(n->in(2), bt);
+
+    // Pattern (1)
+    if (lhs.valid && rhs.valid && lhs.variable == rhs.variable) {
```

Comment:
Would it make sense to add a method to `Multiplication`, let's say `add`. The 3 `if`s here would then be replaced by:
```
Multiplication res = lhs.add(rhs);
if (res.valid()) {
  return res;
}
return find_simple_addition_pattern(n, bt);
```
and the logic from the 3 ifs would be moved to `Multiplication::add`.
What do you think?

Reply:
I added `add(Mulitplication)` and moved `Pattern (1)` to it. However, for pattern `(2)` and `(3)`, the one side is not a `Multiplcation` but a simple node (`(a << CON) + a` and `a + (a << CON)`). Therefore I left the two other `if`s here.